In [14]:
import sqlite3
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Verbinding maken met de database
conn = sqlite3.connect('database.sqlite')

# Een query uitvoeren om de benodigde gegevens op te halen
query = "SELECT * FROM Match;"
match_data = conn.execute(query).fetchall()

# De kolomnamen ophalen
columns_query = "PRAGMA table_info(Match);"
columns = conn.execute(columns_query).fetchall()
column_names = [column[1] for column in columns]

# Het match dataframe maken
import pandas as pd
match_df = pd.DataFrame(match_data, columns=column_names)

# De verbinding met de database sluiten

player_attributes = pd.read_sql_query("SELECT * FROM Player_Attributes", conn)
team_attributes = pd.read_sql_query("SELECT * FROM Team_Attributes", conn)

match_df = pd.read_sql_query("SELECT * FROM Match", conn)

conn.close()

# 1

In [15]:
#Bookmaker: BettingPlanet
odds = {
    'Chelsea': 2.65, 'Manchester City': 4.00, 'Arsenal': 5.40, 'Manchester United': 5.90, 'Liverpool': 29.50,
    'Tottenham': 146.00, 'Everton': 510.00, 'Southampton': 620.00, 'Swansea City': 3001.00, 'Stoke City': 3001.00,
    'West Ham': 3001.00, 'Crystal Palace': 3001.00, 'Newcastle': 3001.00, 'West Brom': 5001.00, 'Aston Villa': 5001.00,
    'Leicester City': 5001.00, 'Sunderland': 5001.00, 'Bournemouth': 7501.00, 'Norwich City': 7501.00, 'Watford': 9751.00
}
class Bookmaker:
    def __init__(self, odds):
        self.odds = odds

    def convert_odds_to_probabilities(self):
        probabilities = {club: 1 / odd for club, odd in self.odds.items()}
        return probabilities

    def print_probabilities(self):
        probabilities = self.convert_odds_to_probabilities()
        print("Probabilities:")
        for club, probability in probabilities.items():
            print(club + ":", probability)



    def predict_result(self):
        probabilities = self.convert_odds_to_probabilities()
        max_probability = max(probabilities, key=probabilities.get)
        return max_probability


bookmaker = Bookmaker(odds)
bookmaker.print_probabilities()

bookmaker = Bookmaker(odds)
predicted_result = bookmaker.predict_result()
print("Predicted Result by Bookmaker:", predicted_result)

Probabilities:
Chelsea: 0.37735849056603776
Manchester City: 0.25
Arsenal: 0.18518518518518517
Manchester United: 0.1694915254237288
Liverpool: 0.03389830508474576
Tottenham: 0.00684931506849315
Everton: 0.00196078431372549
Southampton: 0.0016129032258064516
Swansea City: 0.0003332222592469177
Stoke City: 0.0003332222592469177
West Ham: 0.0003332222592469177
Crystal Palace: 0.0003332222592469177
Newcastle: 0.0003332222592469177
West Brom: 0.0001999600079984003
Aston Villa: 0.0001999600079984003
Leicester City: 0.0001999600079984003
Sunderland: 0.0001999600079984003
Bournemouth: 0.0001333155579256099
Norwich City: 0.0001333155579256099
Watford: 0.00010255358424776946
Predicted Result by Bookmaker: Chelsea


Wat opvalt is dat de teams met lagere odds (zoals Chelsea en Manchester City) hogere kansen hebben, terwijl teams met hogere odds (zoals Tottenham en Watford) lagere kansen hebben. Dit komt doordat de odds de relatieve waarschijnlijkheid van een bepaalde gebeurtenis weergeven volgens de bookmaker.

# 2

In [16]:
odds = {
    'Chelsea': 0.37735849056603776,'Manchester City': 0.25, 'Arsenal': 0.18518518518518517,'Manchester United': 0.1694915254237288,
    'Liverpool': 0.03389830508474576, 'Tottenham': 0.00684931506849315, 'Everton': 0.00196078431372549, 'Southampton': 0.0016129032258064516,
    'Swansea City': 0.0003332222592469177, 'Stoke City': 0.0003332222592469177, 'West Ham': 0.0003332222592469177,
    'Crystal Palace': 0.0003332222592469177,'Newcastle': 0.0003332222592469177,'West Brom': 0.0001999600079984003,
    'Aston Villa': 0.0001999600079984003, 'Leicester City': 0.0001999600079984003, 'Sunderland': 0.0001999600079984003,
    'Bournemouth': 0.0001333155579256099, 'Norwich City': 0.0001333155579256099,'Watford': 0.00010255358424776946
}

def bookmakers_ranking(odds):
    bookmaker_rankings = sorted(odds, key=odds.get, reverse=True)
    print("Predicted Bookmaker Rankings:", bookmaker_rankings)

bookmakers_ranking(odds)

Predicted Bookmaker Rankings: ['Chelsea', 'Manchester City', 'Arsenal', 'Manchester United', 'Liverpool', 'Tottenham', 'Everton', 'Southampton', 'Swansea City', 'Stoke City', 'West Ham', 'Crystal Palace', 'Newcastle', 'West Brom', 'Aston Villa', 'Leicester City', 'Sunderland', 'Bournemouth', 'Norwich City', 'Watford']


2. dit zijn de voorspelde resultaten van de bookmaker met behulp van de kansen. in dit geval zijn de voorspelde resultaten een voorspelling van de ranglijst van de premier league in het seizoen 2015/2016. de bookmaker voorspeld dat Chelsea kampioen wordt en samen met Manchester City, Arsenal en Manchester United volgend seizoen europees gaan spelen in de Champions League. En dat Liverpool en Tottenham volgend seizoen europees gaan spelen in de Europa League en met Everton in de kwalificatie voor de Europa Leaugue. Ook voorspellen ze dat Bournemouth, Norwich City en Watford degraderen en volgend seizoen spelen in de EFL Championship, de 2e compititie van Engeland.

# 3

In [17]:
match_df['result'] = match_df.apply(lambda row: 'win' if row['home_team_goal'] > row['away_team_goal'] else ('defeat' if row['home_team_goal'] < row['away_team_goal'] else 'draw'), axis=1)
pd.set_option('display.max_columns', None)

match_df
import pandas as pd

selected_features = ['chanceCreationShooting','defenceTeamWidth',
                         'defencePressure', 'buildUpPlayPassing', 'defenceAggression', 'chanceCreationCrossing',
                         'chanceCreationPassing',
                         'buildUpPlaySpeed']
df_selected = team_attributes[selected_features]

df_selected['result'] = match_df['result']

display(df_selected)



C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\3039124530.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_selected['result'] = match_df['result']


chanceCreationShooting  defenceTeamWidth  defencePressure  \
0                         55                45               50   
1                         64                54               47   
2                         64                54               47   
3                         70                70               60   
4                         52                52               47   
...                      ...               ...              ...   
1453                      53                53               46   
1454                      50                53               44   
1455                      32                37               44   
1456                      32                37               44   
1457                      32                37               44   

      buildUpPlayPassing  defenceAggression  chanceCreationCrossing  \
0                     50                 55                      65   
1                     56                 44                      63   
2                     54                 44                      63   
3                     70                 70                      70   
4                     52                 47                      48   
...                  ...                ...                     ...   
1453                  52                 48                      48   
1454                  51                 55                      52   
1455                  51                 58                      52   
1456                  51                 58                      52   
1457                  51                 58                      52   

      chanceCreationPassing  buildUpPlaySpeed  result  
0                        60                60    draw  
1                        54                52    draw  
2                        54                47  defeat  
3                        70                70     win  
4                        53                47  defeat  
...                     ...               ...     ...  
1453                     52                52     win  
1454                     47                54  defeat  
1455                     47                54     win  
1456                     47                54    draw  
1457                     47                54  defeat  

[1458 rows x 9 columns]

In [18]:
# Import the required libraries
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# Define the given odds
odds = {
    'Chelsea': 2.65, 'Manchester City': 4.00, 'Arsenal': 5.40, 'Manchester United': 5.90, 'Liverpool': 29.50,
    'Tottenham': 146.00, 'Everton': 510.00, 'Southampton': 620.00, 'Swansea City': 3001.00, 'Stoke City': 3001.00,
    'West Ham': 3001.00, 'Crystal Palace': 3001.00, 'Newcastle': 3001.00, 'West Brom': 5001.00, 'Aston Villa': 5001.00,
    'Leicester City': 5001.00, 'Sunderland': 5001.00, 'Bournemouth': 7501.00, 'Norwich City': 7501.00, 'Watford': 9751.00
}

# Convert the odds to a pandas DataFrame
df = pd.DataFrame.from_dict(odds, orient='index', columns=['Odds'])

# Create a DataFrame to store the team pairs and their labels
team_pairs = pd.DataFrame(columns=['Team1', 'Team2', 'Favorite'])

# Generate team pairs and assign labels
for i in range(len(df)):
    team1 = df.index[i]
    odds1 = df.iloc[i]['Odds']
    for j in range(i+1, len(df)):
        team2 = df.index[j]
        odds2 = df.iloc[j]['Odds']
        if odds1 < odds2:
            favorite = team1
        else:
            favorite = team2
        team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)

# Convert team pairs and labels to numerical representation
team_pairs['Team1'] = team_pairs['Team1'].astype('category').cat.codes
team_pairs['Team2'] = team_pairs['Team2'].astype('category').cat.codes
team_pairs['Favorite'] = team_pairs['Favorite'].astype('category').cat.codes

# Split the data into training and test sets
X = team_pairs[['Team1', 'Team2']].values
y = team_pairs['Favorite'].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Create and train a logistic regression model
logreg_model = LogisticRegression()
logreg_model.fit(X_train, y_train)

# Make predictions using the logistic regression model
y_pred_logreg = logreg_model.predict(X_test)

# Calculate the accuracy of the logistic regression model
accuracy_logreg = accuracy_score(y_test, y_pred_logreg)
print("Logistic regression accuracy:", accuracy_logreg)

# Create and train a decision tree model
dt_model = DecisionTreeClassifier()
dt_model.fit(X_train, y_train)

# Make predictions using the decision tree model
y_pred_dt = dt_model.predict(X_test)

# Calculate the accuracy of the decision tree model
accuracy_dt = accuracy_score(y_test, y_pred_dt)
print("Decision tree accuracy:", accuracy_dt)


C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: Futur

C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: Futur

C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: Futur

C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: Futur

Logistic regression accuracy: 0.3157894736842105
Decision tree accuracy: 0.868421052631579


C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  team_pairs = team_pairs.append({'Team1': team1, 'Team2': team2, 'Favorite': favorite}, ignore_index=True)
C:\Users\jesse\AppData\Local\Temp\ipykernel_12612\1157139725.py:33: Futur

In [7]:
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
import numpy as np

# Stap 1: Data voorbereiding
X_train, X_test, y_train, y_test = train_test_split(df_selected[selected_features], df_selected['result'], test_size=0.2, random_state=42)

# Stap 2: Feature engineering (indien nodig)

# Stap 3: Modelselectie en hyperparameter tuning
logreg_params = {'C': [0.01, 0.1, 1, 10]}
dt_params = {'max_depth': [None, 5, 10, 15]}

# GridSearchCV voor logistische regressie
logreg_grid = GridSearchCV(LogisticRegression(), logreg_params, cv=5)
logreg_grid.fit(X_train, y_train)
best_logreg = logreg_grid.best_estimator_

# GridSearchCV voor decision tree
dt_grid = GridSearchCV(DecisionTreeClassifier(), dt_params, cv=5)
dt_grid.fit(X_train, y_train)
best_dt = dt_grid.best_estimator_

# Stap 4: Model evaluatie
y_pred_logreg = best_logreg.predict(X_test)
accuracy_logreg = accuracy_score(y_test, y_pred_logreg)

y_pred_dt = best_dt.predict(X_test)
accuracy_dt = accuracy_score(y_test, y_pred_dt)

print("Nauwkeurigheid (Logistische regressie):", accuracy_logreg)
print("Nauwkeurigheid (Decision Tree):", accuracy_dt)


C:\Users\jesse\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\jesse\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

Nauwkeurigheid (Logistische regressie): 0.4589041095890411
Nauwkeurigheid (Decision Tree): 0.4657534246575342


We zien hier dat de bookmakers over het algemeen een hogere accuracy hebben dan wij met een Logistic regression accuracy: 0.3157894736842105
 en Decision tree accuracy: 0.868421052631579, tegen over onze 0.4589041095890411 en 0.4657534246575342
 
De reden dat we voor deze modellen hebben gekozen is omdat Logistische Regressie goed werkt bij binaire en multi-class classificatieproblemen zoals deze.
ook is het niet heel moeilijk een gemakkelijk te interpreteren.

Decision Tree:
We hebben gekozen voor dit model, omdat het gebreukt kan worden voor classificatie en regressieproblemen.
Het model kan omgaan met niet lineare relaties, het model kan hier dus op basis van de kenmerken van de teamparen het favoriete team voorspellen. ook bied het model functie-importances wat belangrijk kan zijn om de meest invloedrijke functies voor de voorspellingen te begrijpen.

# 4

Met een nauwkeurigheid van ongeveer 46% (Decision Tree) en 46% (Logistische regressie), zou het model niet betrouwbaar genoeg zijn om te gebruiken bij het nemen van gokbeslissingen. De nauwkeurigheid van het model geeft aan hoe goed het presteert in vergelijking met de werkelijke uitkomsten. Een nauwkeurigheid van 50% zou betekenen dat het model even goed presteert als het willekeurig raden van de uitkomst.

Aangezien de nauwkeurigheid van het model lager is dan 50%, impliceert dit dat het model niet consistent beter presteert dan willekeurig raden.